File: backend/app/api/offers.py
from flask import Blueprint, jsonify, request
from app.models import User, BoardGame, Offer, db

offers_routes = Blueprint('offers', __name__)

# Get offer by Game Owner ID (Offers to Game Owner)
@offers_routes.route('/owner')
def get_offer_by_owner_id():
  user_id = request.args.get('id')
  offers = Offer.query.filter(Offer.owner_id == user_id).order_by(Offer.created_at)
  data = [offer.to_dict() for offer in offers]
  return {"offers": data}, 200

@offers_routes.route('/offer')
def get_offer_by_id():
  offer_id = request.args.get('id')
  offer = Offer.query.filter(Offer.id == offer_id).first()
  return {"offer": offer.to_dict()}

# Get offer by User Offering ID (aka Outstanding Offers)
@offers_routes.route('/offeree/:id')
def get_offer_by_offeree():
  user_id = request.args.get('id')
  offers = Offer.query.filter(Offer.offeree_id == user_id).order_by(Offer.updated_at)
  data = [offer.to_dict() for offer in offers]
  return {"offers": data}, 200

@offers_routes.route('/newoffer', methods=['POST'])
def build_offer():
  try:
    owner_id = request.json.get('ownerId')
    offeree_id = request.json.get('offereeId')
    game_id = request.json.get('gameId')
    offer_buy = bool(request.json.get('offerBuy'))
    offer_trade = bool(request.json.get('offerTrade'))
    offer_borrow = bool(request.json.get('offerBorrow'))
    offer = Offer(
      owner_id=owner_id,
      offeree_id=offeree_id,
      game_id=game_id,
      new_offer=True,
      pending_offer=True,
      offer_buy=offer_buy,
      offer_trade=offer_trade,
      offer_borrow=offer_borrow
    )
    db.session.add(offer)
    db.session.commit()
    return {"Message": "Offer successfully added."}, 200
  except:
    return {"Message": sqlalchemy.exc.DataError}, 500

# View offer by Offer ID
@offers_routes.route('/viewoffer', methods=['PATCH'])
def view_offer():
  try:
    offer_id = request.json.get('offerId')
    offer = Offer.query.filter(Offer.id == offer_id).first()
    if (offer.new_offer == True):
      offer.new_offer = False
    db.session.commit()
    return {"Message": "Offer successfully updated."}
  except:
    return {"Message": "Offer unsuccessfully udpated."}

@offers_routes.route('/delete', methods=['DELETE'])
def delete_offer():
  try:
    offer_id = request.args.get('offerId')
    remove = Offer.query.filter(Offer.id == offer_id).delete()
    db.session.commit()
    return {"Message": "Offer deleted successfully."}
  except:
    return {"Message": "Unable to delete offer."}

File: client/src/store/offers.js
const GET_OFFERS_BY_OWNER_ID = 'offers/get_offers_by_owner_id';
const GET_OFFERS_BY_ID = 'offers/get_offers_by_id';

export const getOffers = (offers) => {
  return {
    type: GET_OFFERS_BY_OWNER_ID,
    offers: offers
  }
}

export const getOffer = (offer) => {
  return {
    type: GET_OFFERS_BY_ID,
    offer: offer
  }
}

export const getOffersByOwnerId = id => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/offers/owner?id=${id}`)
      res.data = await res.json()
      if (res.ok) {
        return dispatch(getOffers(res.data.offers));
      }
      return res
    } catch (err) {
      return console.warn("Error: ", err)
    }
  }
}

export const getOfferById = id => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/offers/offer?id=${id}`)
      res.data = await res.json()
      if (res.ok) {
        console.log(res.data.offer);
        return dispatch(getOffers(res.data.offer));
      }
      return res
    } catch (err) {
      return console.warn("Error: ", err)
    }
  }
}

export const getOffersByOffereeId = id => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/offers/offeree?id=${id}`)
      res.data = await res.json()
      if (res.ok) {
        return dispatch(getOffers(res.data.offers))
      }
      return res
    } catch (err) {
      return console.warn("Error: ", err)
    }
  }
}

export const buildOffer = (ownerId, offereeId, gameId, offerBuy, offerTrade, offerBorrow) => {
  return async dispatch => {
    try {
      console.log(gameId);
      const res = await fetch(`/api/offers/newoffer`, {
        method: 'post',
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ ownerId, offereeId, gameId, offerBuy, offerTrade, offerBorrow })
      })
      res.data = await res.json();
      if (res.ok) {
        dispatch(getOffersByOffereeId(offereeId));
      }
      return res;
    } catch (err) {
      return { "Message": "Coult not generate offer." }
    }
  }
}


export const viewOffer = (userId, offerId) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/offers/viewoffer`, {
        method: "PATCH",
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ offerId })
      })
      res.data = await res.json()
      if (res.ok) {
        dispatch(getOffersByOwnerId(userId))
      }
      return res;
    } catch (err) {
      return { "Message": "Could not update offer." }
    }
  }
}

export const deleteOffer = (userId, offerId) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/offers/delete?offerId=${offerId}`, {
        method: "DELETE",
      })
      res.data = await res.json();
      if (res.ok) {
        dispatch(getOffersByOwnerId(userId))
      }
      return res;
    } catch (err) {
      return { "Message": "Unable to delete offer." }
    }
  }
}

export default function offerReducer(state = {}, action) {
  switch (action.type) {
    case GET_OFFERS_BY_OWNER_ID:
      return { ...state, getOffers: action.offers }
    case GET_OFFERS_BY_ID:
      return {...state, offer: action.offer }
    default:
      return state;
  }
}
